compiler/pipes/final-check.cpp

```diff
@@ -107,6 +109,94 @@ void check_to_array_debug_call(VertexAdaptor<op_func_call> call) {
+
+  for (auto ancestor : klass->get_all_ancestors()) {
+    ancestor->members.for_each([&used_json_keys, ancestor](const ClassMemberInstanceField &field) {
+      auto rename = field.kphp_json_tags->get_rename_value();
+      auto key = rename.empty() ? field.local_name() : rename;
+      bool success = used_json_keys.emplace(key).second;
+      kphp_error_return(success, fmt_format("@kphp-json field '{}' met twice, first time in class '{}'", key, ancestor->name));
+    });
+  }
+}
+
+void check_json_tag_correctness(ClassPtr klass) noexcept {
+  klass->members.for_each([](const ClassMemberInstanceField &field) {
+    const auto type = tinf::get_type(field.var)->ptype();
+    if (field.kphp_json_tags->has_array_as_hashmap()) {
+      kphp_error_return(type == tp_array,
+        fmt_format("@kphp-json array_as_hashmap tag are only allowed above array type, got above {} field", field.local_name()));
+    }
+    if (field.kphp_json_tags->is_raw_string()) {
+      kphp_error_return(type == tp_string,
+        fmt_format("@kphp-json raw_string tag are only allowed above string type, got above {} field", field.local_name()));
+    }
+  });
+}
+
+void store_json_encoder(ClassPtr klass, ClassPtr encoder) noexcept {
+  if (klass->parent_class) {
+    store_json_encoder(klass->parent_class, encoder);
+  }
+
+  AutoLocker<Lockable *> lock{&(*klass)};
+  auto it = std::find(klass->json_encoders.begin(), klass->json_encoders.end(), encoder);
+  if (it == klass->json_encoders.end()) {
+    klass->json_encoders.push_front(encoder);
+  }
+
+  klass->members.for_each([klass, encoder](ClassMemberInstanceField &field) {
+    std::unordered_set<ClassPtr> sub_classes;
+    field.var->tinf_node.get_type()->get_all_class_types_inside(sub_classes);
+    for (auto sub_klass : sub_classes) {
+      if (sub_klass != klass) {
+        store_json_encoder(sub_klass, encoder);
+      }
+    }
+  });
+}
+
+ClassPtr extract_json_encoder_from_call(VertexAdaptor<op_func_call> call) noexcept {
+  auto first_arg = call->args().front();
+  kphp_error(first_arg, "absent first arg");
+  auto first_arg_var = first_arg.try_as<op_var>();
+  kphp_error(first_arg_var, "first arg shoule be op_var type");
+  auto v_encoder = first_arg_var->var_id->init_val.try_as<op_string>();
+  kphp_error(v_encoder, "absent init value");
+  auto encoder = G->get_class(v_encoder->str_val);
+  kphp_error(encoder, "couldn't finc encoder class");
+  return encoder;
+}
+
+void check_to_json_impl_call(VertexAdaptor<op_func_call> call) noexcept {
+  kphp_error_return(call->args().size() >= 2, "incorrect args count in to_json_impl call");
+  const auto *type = tinf::get_type(call->args()[1]);
+  kphp_error_return(type->ptype() == tp_Class, "second argument of to_json_impl should be a class type");
+  const auto klass = type->class_type();
+  kphp_error(!klass->is_ffi_cdata(), "called to_json_impl with CData");
+  check_json_fields_duplication(klass);
+  check_json_tag_correctness(klass);
```

Comment:
I'd suggest to call `check_json_fields_duplication` and `check_json_tag_correctness` not on a func call, but once per class: in FinalCheckPass::on_start, inside already existing `if (type == func_class_holder)`
BTW, also insert
```
    stage::set_line(current_function->class_id->location_line_num);
```
into that if, to get better error messages.

Reply:
Actually there is no simple condition in class that allow to recognize him as jsonable. So there are two options here: add one more json flag to class (which complicate code), or determine whether class jsonable by traverse its members (need to do that for all classes, not only with `to_json(class_instance<A>)` which is slower). I dislike both of them, let's leave as is?

Reply:
`json_encoders` filled after call to `on_start`, so let me leave this as is